### PR TITLE
Store API tokens encrypted in MongoDB

### PR DESCRIPTION
**Pull request by thll**
## Motivation and Context

At the moment, API tokens are stored in plaintext. If we still want to support that they can be copied to the clipboard, we can't hash them. But we can at least try to encrypt them with symmetric encryption.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by thll** [comment]
If we want to stick to the current approach, we could have a look at https://github.com/codahale/aes-gcm-siv. I got it working with a quick-and-dirty prototype but I'm not sure about the security implications yet.

**Comment by bernd** [comment]
> If we want to stick to the current approach, we could have a look at https://github.com/codahale/aes-gcm-siv. I got it working with a quick-and-dirty prototype but I'm not sure about the security implications yet.

@thll Agreed, I came up with the same conclusion during my research. AES SIV sounds the way to go for our use case. It has security implications, but at least it's standardized and with https://github.com/cryptomator/siv-mode there seems to be an actively maintained Java implementation. (Coda's repository is archived)

File: graylog2-server/src/main/java/org/graylog2/rest/models/users/responses/Token.java [hunk 1]
```diff
@@ -23,10 +23,16 @@
...
 import org.joda.time.DateTime;
 
+import javax.annotation.Nullable;
+
 @JsonAutoDetect
 @AutoValue
 @WithBeanGetter
 public abstract class Token {
+    @JsonProperty
+    @Nullable
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/rest/models/users/responses/Token.java]: :heavy_check_mark:

File: graylog2-server/src/main/java/org/graylog2/security/token/AccessTokenServiceImpl.java [hunk 2]
```diff
@@ -60,9 +67,19 @@ public AccessToken load(String token) {
...
-        final Object id = tokenObject.get("_id");
-        return new AccessTokenImpl((ObjectId) id, tokenObject.toMap());
+        return fromDBObject(objects.get(0));
+    }
+
+    @Nullable
+    @Override
+    public AccessToken loadById(String id) {
+        try {
+            return fromDBObject(get(AccessTokenImpl.class, id));
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/security/token/AccessTokenServiceImpl.java]: :heavy_check_mark:

File: graylog2-server/src/test/java/org/graylog2/security/AccessTokenServiceImplTest.java [hunk 3]
```diff
@@ -30,6 +33,9 @@
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 public class AccessTokenServiceImplTest {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/test/java/org/graylog2/security/AccessTokenServiceImplTest.java]: Agreed, tests have been added. :heavy_check_mark:

File: graylog2-server/src/main/java/org/graylog2/security/AESTools.java [hunk 4]
```diff
@@ -55,4 +61,61 @@ public static String decrypt(String cipherText, String encryptionKey, String sal
...
+                    Hex.decode(cipherText.getBytes(UTF_8))
+            );
+            return new String(plainBytes, UTF_8);
+        } catch (Exception e) {
+            LOG.error("Couldn't decrypt value", e);
+        }
+        return null;
+    }
+
+    private static void validateTextAndEntryptionKey(String text, byte[] encryptionKey) {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/security/AESTools.java]: :heavy_check_mark:

File: graylog2-server/src/main/java/org/graylog2/security/AccessToken.java [hunk 5]
```diff
@@ -23,6 +23,24 @@
  * @author Dennis Oelkers <dennis@torch.sh>
  */
 public interface AccessToken extends Persisted {
+    enum Type {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/security/AccessToken.java]: Agreed! :heavy_check_mark:

File: graylog2-server/src/main/java/org/graylog2/security/token/AccessTokenImpl.java [hunk 6]
```diff
@@ -81,6 +83,7 @@ public String getToken() {
     @Override
     public void setToken(String token) {
         fields.put(TOKEN, token);
+        fields.put(TOKEN_TYPE, Type.defaultType().getIntValue());
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/security/token/AccessTokenImpl.java]: :heavy_check_mark:

File: graylog2-server/src/main/java/org/graylog2/security/token/AccessTokenServiceImpl.java [hunk 7]
```diff
@@ -14,43 +14,50 @@
...
+import javax.annotation.Nullable;
 import javax.inject.Inject;
 import java.math.BigInteger;
 import java.security.SecureRandom;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
 
 public class AccessTokenServiceImpl extends PersistedServiceImpl implements AccessTokenService {
```
[the remark this answers is not in the document]
Comment (bernd) [reply graylog2-server/src/main/java/org/graylog2/security/token/AccessTokenServiceImpl.java]: :heavy_check_mark:

**Pull request merged**